lib/scholar/linear/bayesian_ridge_regression.ex

```diff
@@ -0,0 +1,320 @@
+      `:lambda` must be a non-negative float i.e. in [0, inf).
+      Defaults to 1.
+      """
+    ],
+    alpha_1: [
+      type: {:custom, Scholar.Options, :non_negative_number, []},
+      default: 1.0e-6,
+      doc: ~S"""
+      Hyper-parameter : shape parameter for the Gamma distribution prior      
+      over the alpha parameter.
+      """
+    ],
+    alpha_2: [
+      type: {:custom, Scholar.Options, :non_negative_number, []},
+      default: 1.0e-6,
+      doc: ~S"""
+      Hyper-parameter : inverse scale (rate) parameter for the Gamma distribution prior
+      over the alpha parameter.
+      """
+    ],
+    lambda_1: [
+      type: {:custom, Scholar.Options, :non_negative_number, []},
+      default: 1.0e-6,
+      doc: ~S"""
+      Hyper-parameter : shape parameter for the Gamma distribution prior
+      over the lambda parameter.            
+      """
+    ],
+    lambda_2: [
+      type: {:custom, Scholar.Options, :non_negative_number, []},
+      default: 1.0e-6,
+      doc: ~S"""
+      Hyper-parameter : inverse scale (rate) parameter for the Gamma distribution prior      
+      over the lambda parameter.            
+      """
+    ],
+    eps: [
+      type: :float,
+      default: 1.0e-8,
+      doc:
+        "The convergence tolerance. When `Nx.sum(Nx.abs(coef - coef_new)) < :eps`, the algorithm is considered to have converged."
+    ]
+  ]
+
+  @opts_schema NimbleOptions.new!(opts)
+  deftransform fit(x, y, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+
+    opts =
+      [
+        sample_weights_flag: opts[:sample_weights] != nil
+      ] ++
+        opts
+
+    {sample_weights, opts} = Keyword.pop(opts, :sample_weights, 1.0)
+    x_type = to_float_type(x)
+
+    sample_weights =
+      if Nx.is_tensor(sample_weights),
+        do: Nx.as_type(sample_weights, x_type),
+        else: Nx.tensor(sample_weights, type: x_type)
+
+    # handle vector types
+    # handle default alpha value, add eps to avoid division by 0
+    eps = Nx.Constants.smallest_positive_normal(x_type)
+    default_alpha = Nx.divide(1, Nx.add(Nx.variance(x), eps))
+    alpha = Keyword.get(opts, :alpha_init, default_alpha)
+    alpha = Nx.tensor(alpha, type: x_type)
+    opts = Keyword.put(opts, :alpha_init, alpha)
+
+    {lambda, opts} = Keyword.pop!(opts, :lambda_init)
+    lambda = Nx.tensor(lambda, type: x_type)
+    opts = Keyword.put(opts, :lambda_init, lambda)
+    zeros_list = for k <- 0..opts[:iterations], do: 0
+    scores = Nx.tensor(zeros_list, type: x_type)
+
+    {coefficients, intercept, alpha, lambda, rmse, iterations, has_converged, scores, sigma} =
+      fit_n(x, y, sample_weights, scores, opts)
+    iterations = Nx.to_number(iterations)
+    scores = scores
+    |> Nx.to_list()
+    |> Enum.take(iterations)
+
+    if Nx.to_number(has_converged) == 1 do
+      IO.puts("Convergence after #{Nx.to_number(iterations)} iterations")
+    end
```

Comment:
You really can't call this here. You have to assume you never can really read the tensor values, even inside `deftransform`. You can call: `Nx.Defn.jit(&BayesianRidgeRegression.fit/3).(x, y, opts)` and you will see these operations will fail. :) You can even add this as a test to check the operation is fully jittable. This means you can't do `Nx.to_list() |> Enum.take(iterations)` either. Whoever calls `fit`has to do that instead.